### PR TITLE
Faster retraction and inverse retractions for Stiefel

### PR DESCRIPTION
**Pull request by mateuszbaran**
Inverse QR retraction is now much faster for k=1 and k=2 :slightly_smiling_face:.

### CODE REVIEW

**Comment by kellertuer** [comment]
Cool. I will look at the details the next days.
Nice, are the missing parts in code coverage covered but just not detected?

Is it maybe time to also do a benchmark ci for these things?

**Comment by mateuszbaran** [comment]
Thanks! Missing coverage was actually missing, I've pushed a new commit with some tests. I'm actually not even sure why CI thinks the k>2 case in inverse QR retraction is covered, I don't know which test covers it.

Yes, benchmark CI would be great. I'll think about adding it.

**Comment by mateuszbaran** [comment]
I've also added faster distance on SPD manifold with linear-affine metric. It's a bit more explicit because Julia had trouble with efficiently figuring out that it's a generalized symmetric positive definite eigenproblem. It would be even faster if StaticArrays had right matrix division by triangular matrices (it only has the left variant) but it's a problem for another day.

File: src/manifolds/Stiefel.jl
Comment (kellertuer) [inline src/manifolds/Stiefel.jl:198]: And we can't do this via dispatch on p/q because for A you want to promote the type first, right?

It's just a matter of style, that I would prefer shorter functions and dispatch if possible.

**Review by kellertuer — APPROVED** [review]
With my only remark on the length of the function – I'll let you decide whether thats possible but already approve this PR.

File: src/manifolds/Stiefel.jl
Comment (mateuszbaran) [reply src/manifolds/Stiefel.jl:198]: I can split this into more functions, yes.

Comment (mateuszbaran) [reply src/manifolds/Stiefel.jl]: Now we have dispatch and lots of small functions instead :smile: .

**Comment by mateuszbaran** [comment]
I will tag a new minor version after this is merged.

**Comment by mateuszbaran** [comment]
These optimizations require StaticArrays master unfortunately. Would you be OK with disabling these tests until next version of StaticArrays is released?

**Comment by kellertuer** [comment]
Can we until then document that as a note in the docstring? We can remove that note once it works (i.e. when we enable the tests)?

**Comment by mateuszbaran** [comment]
Hm... I'm thinking about either making it work with StaticArrays 0.12 in some way or putting the note.

**Comment by mateuszbaran** [comment]
I've disabled the optimization that requires StaticArrays master for now (it's commented out). I'm going to merge & tag when CI passes.

**Pull request merged**
